packages/framework-core/src/services/event-store.ts

```diff
@@ -44,40 +49,7 @@ export class EventStore {
     }
   }
 
-  public async calculateAndStoreEntitySnapshot(
-    entityName: string,
-    entityID: UUID,
-    pendingEnvelopes: Array<EventEnvelope>
-  ): Promise<EventEnvelope | null> {
-    const logger = getLogger(this.config, 'EventStore#calculateAndStoreEntitySnapshot')
-    logger.debug('Processing events: ', pendingEnvelopes)
-    logger.debug(`Fetching snapshot for entity ${entityName} with ID ${entityID}`)
-    const latestSnapshotEnvelope = await this.loadLatestSnapshot(entityName, entityID)
-
-    logger.debug(
-      `[EventStore#calculateAndStoreEntitySnapshot] Looking for the reducer for entity ${entityName} with ID ${entityID}`
-    )
-    let newEntitySnapshot = latestSnapshotEnvelope
-    for (const pendingEvent of pendingEnvelopes) {
-      newEntitySnapshot = await this.entityReducer(newEntitySnapshot, pendingEvent)
-    }
-
-    logger.debug(
-      `[EventStore#calculateAndStoreEntitySnapshot] Reduced new snapshot for entity ${entityName} with ID ${entityID}: `,
-      newEntitySnapshot
-    )
-
-    if (!newEntitySnapshot) {
-      logger.debug('New entity snapshot is null. Returning old one (which can also be null)')
-      return latestSnapshotEnvelope
-    }
-
-    await this.storeSnapshot(newEntitySnapshot)
-
-    return newEntitySnapshot
-  }
-
-  private async storeSnapshot(snapshot: EventEnvelope): Promise<void> {
+  public async storeSnapshot(snapshot: EventEnvelope): Promise<void> {
```

Comment:
I think that we should have a unique high-level method with a signature like the following:

```typescript
public async fetchEntitySnapshot(entityName: string, entityID: UUID, timestamp?: string): Promise<EventEnvelope | null> {
  /* 
  1. If a timestamp is provided, we get the snapshot at the specified time (enabling the long awaited time-traveling feature 😎 ). If not, you get the latest known state.
  2. It looks for the latest snapshot.
    2.1. If not found, it looks for the events since the beginning of time
    2.2. If no events found, returns null
  3. If snapshot is found, it looks for events after that snapshot
    3.1. If no events, the snapshot is returned right away without caching it
    3.2. If there are events, it reduces all pending events to generate a new snapshot and stores the newly calculated snapshot before returning it.
  */
}
```

Notice that for this to work, it's important to guarantee that the snapshots and the events related to the same entity are being stored in the same shard of the database, otherwise, this algorithm would cause consistency issues at scale.

Also, I think that we should ignore the events that we already have in the event dispatcher, and use them only as a signal that a specific entity needs to be refreshed. So, I'd remove the `pendingEnvelopes` parameter. This makes the algorithm more resilient to event gaps if for some reason a page of events is not properly processed, and in case of errors, failing events are going to be retried until a bug is solved. Making this extra I/O makes event processing slower, but this process is run *after* event ingestion and I think that we gain significantly more resilience.

Caching snapshots every time they're calculated optimizes consecutive reads. With the current algorithm, we have seen how when event handlers need to load the current state of an entity using `Booster.entity`, the snapshot calculation runs several times. It's also safe to store on read, because when the sharding strategy guarantees ordering, snapshot calculation is idempotent: storing two snapshots that are calculated in parallel would have no overall effect, as both calculations will always produce the same exact result. Also, given that the ID of the snapshot is tied to the last event read, both calculations will end up having the same ID, making the second one of them overwrite the first one with the same value, ending with a single record in the database anyway.

I know that having operations that look like read but are also updating are not much functional, but in this case I don't see caching as a side effect, but as a way to optimize the overall algorithm. The algorithm would work exactly the same way if we didn't cache the results, it would just be slower.

Reply:
Sounds good, I'd merge the 3 PRs first and then refactor into this algorithm you're describing in order to avoid merging conflicts in the other PRs given that this is a major restructure